src/riak_core_gossip.erl

```diff
@@ -367,73 +371,291 @@ remove_from_cluster(Ring, ExitingNode) ->
                 %% just so rebalance doesn't include exiting node
-                Members = riak_core_ring:claiming_members(Ring),
-                Other = hd(lists:delete(ExitingNode, Members)),
-                TempRing = lists:foldl(
-                             fun({I,N}, R) when N == ExitingNode ->
-                                     riak_core_ring:transfer_node(I, Other, R);
-                                (_, R) -> R
-                             end,
-                             Ring,
-                             AllOwners),
-                riak_core_claim:claim_rebalance_n(TempRing, Other)
+                HN = hd(lists:delete(ExitingNode, Members)),
+                TempRing =
+                    lists:foldl(fun({I,N}, R) when N == ExitingNode ->
+                                        riak_core_ring:transfer_node(I, HN, R);
+                                    (_, R) ->
+                                        R
+                                end,
+                                Ring,
+                                Owners),
+                riak_core_claim:sequential_claim(TempRing, HN)
         end,
     ExitRing.
 
-attempt_simple_transfer(Seed, Ring, Owners, ExitingNode) ->
-    TargetN = app_helper:get_env(riak_core, target_n_val),
-    attempt_simple_transfer(Seed, Ring, Owners,
-                            TargetN,
-                            ExitingNode, 0,
-                            [{O,-TargetN} || O <- riak_core_ring:claiming_members(Ring),
-                                             O /= ExitingNode]).
-attempt_simple_transfer(Seed, Ring, [{P, Exit}|Rest], TargetN, Exit, Idx, Last) ->
-    %% handoff
-    case [ N || {N, I} <- Last, Idx-I >= TargetN ] of
+-ifdef(TEST).
+-type transfer_ring() :: [{integer(), term()}].
+-else.
+-type transfer_ring() :: riak_core_ring:riak_core_ring().
+-endif.
+
+%% @doc Simple transfer of leaving node's vnodes to safe place
+%% Where safe place is any node that satisfies target_n_val for that vnode -
+%% but with a preference to transfer to a node that has a lower number of 
+%% vnodes currently allocated.
+%% If safe places cannot be found for all vnodes returns `target_n_fail`
+%% Simple transfer is not location aware, but generally this wll be an initial
+%% phase of a plan, and hence a temporary home - so location awareness is not
+%% necessary.
+%% `riak_core.full_rebalance_onleave = true` may be used to avoid this step,
+%% although this may result in a large number of transfers
+-spec attempt_simple_transfer(transfer_ring(),
+                                term(),
+                                random:ran(),
+                                [{integer(), term()}],
+                                [term()]) ->
+                                    {ok, transfer_ring()}|
+                                        target_n_fail|
+                                        force_rebalance.
+attempt_simple_transfer(Ring, ExitingNode, Seed, Owners, Members) ->
```

Comment:
The function relies on the fact that the list of owners is sorted. However, the resulting Ring is not necessarily sorted.